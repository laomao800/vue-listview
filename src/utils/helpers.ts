import { VNode } from 'vue'
import isPlainObject from 'lodash/isPlainObject'
import isEmpty from 'lodash/isEmpty'
import isFunction from 'lodash/isFunction'
import get from 'lodash/get'
import { FilterField } from '~/types'

/**
 * 判断 node 是否为 Vue 内部的 VNode 类型
 */
export function isVNode(node: any): node is VNode {
  return (
    node !== null &&
    typeof node === 'object' &&
    hasOwn(node, 'componentOptions')
  )
}

/**
 * 根据映射表设置的结构提取数据并返回一个新对象
 *
 * dataMapping({
 *   result: {
 *     results: [1, 2, 3]
 *   }
 * }, {
 *   items: 'result.results'
 * })
 *
 * -> { items: [1, 2, 3] }
 *
 * @param {Object} data 目标数据
 * @param {Object} dataMap 映射表
 */
export function dataMapping(
  data: Record<string, any> = {},
  dataMap: Record<string, any> = {}
) {
  const result: Record<string, any> = {}

  if (isPlainObject(dataMap)) {
    Object.keys(dataMap).forEach((key) => {
      try {
        const dataKey = key.toString()
        const dataValue = get(data, dataMap[key])
        result[dataKey] = dataValue
      } catch (e) {}
    })
  }

  return result
}

/**
 * 判断值是否为搜索栏内合法的值，通过验证的值才可继续作为参数随请求提交
 */
export function isValidFieldValue(val: any): val is FilterField {
  return !(
    val === null ||
    val === undefined ||
    val === '' ||
    ((Array.isArray(val) || isPlainObject(val)) && isEmpty(val))
  )
}

/**
 * 判断一个值是否有效的搜索栏 field 配置项
 */
export function isValidFieldConfig(field: any) {
  return (
    field &&
    (hasOwn(field, 'type') ||
      isFunction(field.render) ||
      isFunction(field) ||
      isVNode(field))
  )
}

export function nodeParents(node: Element, selector: string) {
  const allMatchs = Array.from(document.querySelectorAll(selector))
  if (allMatchs.length === 0) {
    return null
  }
  function find(curNode: Element): Element | null {
    const parentNode = curNode.parentNode as Element | null
    if (parentNode) {
      if (allMatchs.includes(parentNode)) {
        return parentNode
      } else if (parentNode.parentNode) {
        return find(parentNode)
      }
    }
    return null
  }
  return find(node)
}

export function isPromise(obj: any) {
  return (
    !!obj &&
    (typeof obj === 'object' || typeof obj === 'function') &&
    typeof obj.then === 'function'
  )
}

export function hasOwn(obj: Record<string, unknown>, key: string) {
  return Object.prototype.hasOwnProperty.call(obj, key)
}

export function isDef(val: any) {
  return val !== undefined
}

export function hasRenderFn<T>(item: any): item is T {
  return isPlainObject(item) && isFunction(item.render)
}

export function ensurePromise<T>(data: T) {
  return isPromise(data) ? data : Promise.resolve(data)
}

export function noop() {
  // noop
}

export function resolveOptions(
  optionsConfig: any,
  done: (options: any[]) => void
) {
  let optionsPromise = null
  if (optionsConfig) {
    if (Array.isArray(optionsConfig)) {
      optionsPromise = Promise.resolve(optionsConfig)
    } else if (isFunction(optionsConfig)) {
      optionsPromise = ensurePromise(optionsConfig(done))
    } else if (isPromise(optionsConfig)) {
      optionsPromise = optionsConfig
    }
  }
  return optionsPromise
}
