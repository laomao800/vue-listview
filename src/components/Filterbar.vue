<template>
  <el-form
    :inline="true"
    size="small"
    class="lv__filterbar"
    @submit.native.prevent
    @keydown.native.enter="handleSubmit"
  >
    <div v-if="$slots['filterbar-top']" class="lv__filterbar-top">
      <slot name="filterbar-top" />
    </div>

    <div class="lv__filterbar-main">
      <div v-if="$slots['filterbar-left']" class="lv__filterbar-left">
        <slot name="filterbar-left" />
      </div>

      <div
        v-if="isShowFilterButtons || isShowFilterSubmit || isShowFilterFields"
        :class="[
          'lv__filterbar-inner',
          { 'lv__filterbar-inner--fold': isFold },
        ]"
      >
        <!-- 提交、重置按钮区域 -->
        <div
          v-if="isShowFilterSubmit"
          ref="action"
          :class="[
            'lv__filterbar-action',
            {
              'lv__filterbar-action--nomore': isNoMore,
              'lv__filterbar-action--onleft': isNoneFields,
            },
          ]"
        >
          <el-form-item
            :style="{ transform: `translateX(${searchBtnOffset}px)` }"
            class="lv__filterbar-action-submit"
          >
            <slot name="prepend-submit" />
            <el-button
              v-if="isShowSearchButton"
              v-bind="searchButton"
              @click="handleFilterSearch"
            >
              {{ searchButton.text }}
            </el-button>
            <el-button
              v-if="isShowResetButton"
              v-bind="resetButton"
              @click="handleFilterReset"
            >
              {{ resetButton.text }}
            </el-button>
            <slot name="append-submit" />
          </el-form-item>
          <div class="lv__filterbar-action-ext">
            <slot name="prepend-more" />
            <el-button
              icon="el-icon-caret-top"
              type="primary"
              class="lv__filterbar-action-more"
              @click="toggleFilterbar"
            />
            <slot name="append-more" />
          </div>
        </div>

        <!-- 操作按钮区域 -->
        <FilterbarButtons
          v-if="isShowFilterButtons"
          :buttons="filterButtons"
          class="lv__filterbar-buttons"
        />

        <!-- 搜索栏控件区域 -->
        <FilterbarFields
          v-if="isShowFilterFields"
          ref="FilterbarFields"
          :fields="filterFields"
          class="lv__filterbar-fields"
        />
      </div>

      <div v-if="$slots['filterbar-right']" class="lv__filterbar-right">
        <slot name="filterbar-right" />
      </div>
    </div>

    <div v-if="$slots['filterbar-bottom']" class="lv__filterbar-bottom">
      <slot name="filterbar-bottom" />
    </div>
  </el-form>
</template>

<script lang="tsx">
import Vue, { PropType } from 'vue'
import { hasOwn } from '@/utils'
import FilterbarButtons from './FilterbarButtons.vue'
import FilterbarFields from './FilterbarFields.vue'
import storeProviderMixin from '@/mixins/storeProviderMixin'
import { FilterButton, FilterField } from '~/types'

export default Vue.extend({
  name: 'Filterbar',

  mixins: [storeProviderMixin],

  inheritAttrs: false,

  components: {
    FilterbarButtons,
    FilterbarFields,
  },

  props: {
    filterButtons: {
      type: Array as PropType<FilterButton[]>,
      default: /* istanbul ignore next */ () => [],
    },
    filterFields: {
      type: Array as PropType<FilterField[]>,
      default: /* istanbul ignore next */ () => [],
    },
    searchButton: {
      type: [Object, Boolean],
      default: () => ({
        text: '搜索',
        icon: 'el-icon-search',
        type: 'primary',
      }),
    },
    resetButton: {
      type: [Object, Boolean],
      default: () => ({
        text: '重置',
        icon: '',
        type: 'default',
      }),
    },
  },

  data() {
    return {
      isFold: true,
      isNoMore: true,
      topRightItemIndex: -1,
      actionOffsetLeft: 0,
      searchBtnOffset: 0,
    }
  },

  computed: {
    isNoneFields(): boolean {
      return this.filterFields.length === 0
    },
    isShowFilterButtons(): boolean {
      return this.filterButtons.length > 0
    },
    isShowFilterFields(): boolean {
      return this.filterFields.length > 0
    },
    isShowSearchButton(): boolean {
      return !!this.searchButton
    },
    isShowResetButton(): boolean {
      return !!this.resetButton
    },
    isShowFilterSubmit(): boolean {
      return !!(
        this.isShowSearchButton ||
        this.isShowResetButton ||
        this.$slots['prepend-filterbar-submit'] ||
        this.$slots['append-filterbar-submit']
      )
    },
    allFieldsVm(): Vue[] {
      try {
        return (this as any).$refs['FilterbarFields'].$refs['field'] || []
      } catch (error) {
        return []
      }
    },
  },

  watch: {
    isShowSearchButton: 'updateLayout',
    isShowResetButton: 'updateLayout',
  },

  methods: {
    handleSubmit() {
      this.lvStore.pressEnterSearch && this.handleFilterSearch()
    },

    handleFilterSearch() {
      this.$rootEmitProxy('filter-submit')
      this.lvStore.search()
    },

    handleFilterReset() {
      const filterModel = this.lvStore.filterModel
      const _resetField = (field: FilterField) => {
        const name = field.model
        if (name && hasOwn(filterModel, name)) {
          const value = filterModel[name]
          if (Array.isArray(value)) {
            this.$set(filterModel, name, [])
          } else {
            this.$set(filterModel, name, undefined)
          }
        }
      }
      this.filterFields.forEach((field) => {
        if (Array.isArray(field)) {
          field.forEach(_resetField)
        } else {
          _resetField(field)
        }
      })
      this.$rootEmitProxy('filter-reset')
    },

    toggleFilterbar() {
      this.isFold = !this.isFold
      this.$emit('fold-change', this.isFold)
    },

    updateLayout() {
      // updateTopRightItemIndex 影响 isNoMore 按钮显示，需计算后再执行按钮偏移量计算
      this.updateTopRightItemIndex()
      this.$nextTick().then(() => {
        this.updateActionOffset()
        this.updateBtnOffset()
      })
    },

    getAllFieldsDom() {
      try {
        return (this as any).$refs['FilterbarFields'].$refs['field'] || []
      } catch (error) {
        return []
      }
    },

    updateTopRightItemIndex() {
      let lastFilterIndex = -1
      const allFields = this.allFieldsVm
      if (allFields.length > 0) {
        const $action = this.$refs.action as Element
        let lastFilterTop = $action.getBoundingClientRect().top
        for (let i = 0; i < allFields.length; i++) {
          const curItemTop = allFields[i].$el.getBoundingClientRect().top
          if (lastFilterTop < curItemTop) {
            break
          }
          lastFilterIndex = i
        }
      }
      this.topRightItemIndex = lastFilterIndex
      this.isNoMore = this.topRightItemIndex === this.filterFields.length - 1
    },

    updateBtnOffset() {
      let offset = 0
      if (this.topRightItemIndex >= 0) {
        const lastItem = this.allFieldsVm[this.topRightItemIndex].$el
        const { left, width } = lastItem.getBoundingClientRect()
        offset = left + width - this.actionOffsetLeft + 10
        offset = Math.min(0, offset)
      }
      this.searchBtnOffset = Math.floor(offset)
    },

    updateActionOffset() {
      const $action = this.$refs.action as Element
      if ($action) {
        this.actionOffsetLeft = $action.getBoundingClientRect().left
      }
    },
  },
})
</script>

<style lang="less">
@filter-gap-size: 10px;

.lv__filterbar-top {
  margin-bottom: @filter-gap-size;
}
.lv__filterbar-bottom {
  margin-bottom: @filter-gap-size;
}
.lv__filterbar-left {
  margin-right: @filter-gap-size;
  margin-bottom: @filter-gap-size;
}
.lv__filterbar-right {
  margin-left: @filter-gap-size;
  margin-bottom: @filter-gap-size;
}
.lv__filterbar-main {
  display: flex;
}

.lv__filterbar {
  .el-button {
    [class*=' el-icon-'],
    [class^='el-icon-'] {
      width: 1em;
    }

    height: 32px;
    padding-top: 0;
    padding-bottom: 0;
    line-height: 1;
    vertical-align: top;
  }
  .el-input-group__append,
  .el-input-group__prepend {
    padding: 0 10px;
  }
  .el-form-item__label {
    padding: 0 0 0 2px;
    margin-right: -4px;
  }
  .el-form-item {
    margin: 0;
  }
}

.lv__filterbar {
  &-buttons {
    float: left;
    margin-right: 0;

    .el-button + .el-dropdown,
    .el-dropdown + .el-button,
    .el-dropdown + .el-dropdown,
    .el-form-item__content > * {
      display: inline-block;
      margin-right: @filter-gap-size;
      margin-left: 0;
    }
  }

  &-action {
    display: flex;
    float: right;
    margin: 0;
    margin-bottom: 10px;

    &-submit .el-form-item__content,
    &-ext {
      display: flex;
      & > * {
        display: inline-block;
        transition: inherit;
      }
      & > *:not(:nth-child(1)) {
        margin-left: 10px;
      }
    }

    &-ext {
      // display: flex;
      float: right;
      margin-left: 10px;
    }

    &-more {
      width: 40px;
      padding: 0;
      line-height: 30px;
      transition: none;
    }

    &--nomore {
      .lv__filterbar-action-more {
        display: none;
      }
    }

    &--onleft {
      float: none;
      display: inline-block;
    }
  }
}

.lv__filterbar-inner {
  flex: 1;
  padding-top: @filter-gap-size;
  margin-top: -@filter-gap-size;

  &::after {
    display: table;
    clear: both;
    content: '';
  }

  &--fold {
    box-sizing: content-box;
    height: 32px;
    overflow: hidden;
    margin-bottom: @filter-gap-size;

    .lv__filterbar-action-more {
      transform: rotate(180deg);
    }
  }
}
</style>
