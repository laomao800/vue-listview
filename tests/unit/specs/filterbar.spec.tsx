import Vue from 'vue'
import { mount } from '@vue/test-utils'
import { wait } from '../helpers'
import Filterbar from '@/components/Filterbar.vue'

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const h = new Vue().$createElement
const DATE1 = new Date('2021/01/01 09:30:00')
const DATE2 = new Date('2021/06/01 09:30:00')

describe('Filter buttons', () => {
  it('normal', async () => {
    const filterButtons: any = [
      { text: 'default' },
      { type: 'text', text: 'text' },
      { type: 'primary', icon: 'el-icon-edit', text: 'primary' },
      { type: 'success', icon: 'el-icon-check', text: 'success' },
      { type: 'info', icon: 'el-icon-message', text: 'info' },
      { type: 'warning', icon: 'el-icon-star-off', text: 'warning' },
      { type: 'danger', icon: 'el-icon-delete', text: 'danger' },
    ]
    const wrapper = mount(Filterbar, { propsData: { filterButtons } })
    const buttons = wrapper
      .findComponent({ name: 'FilterbarButtons' })
      .findAllComponents({ name: 'ElButton' })
    expect(wrapper.html()).toMatchSnapshot()
    expect(buttons.length).toBe(filterButtons.length)
  })

  it('dropdown', () => {
    const wrapper = mount(Filterbar, {
      propsData: {
        filterButtons: [
          {
            type: 'primary',
            text: '下拉按钮',
            children: [
              { icon: 'el-icon-circle-plus-outline', text: '菜单1' },
              { icon: 'el-icon-remove-outline', text: '菜单2' },
            ],
          },
        ],
      },
    })
    const buttons = wrapper
      .findComponent({ name: 'FilterbarButtons' })
      .findAllComponents({ name: 'ElButton' })
    const dropdownItem = wrapper.findAllComponents({
      name: 'ElDropdownItem',
    })
    expect(buttons.length).toBe(1)
    expect(dropdownItem.length).toBe(2)
  })

  it('click', () => {
    const mockFn1 = jest.fn()
    const mockFn2 = jest.fn()
    const mockFn3 = jest.fn()
    const filterButtons = [
      { text: 'btn', click: mockFn1 },
      {
        text: 'btn',
        children: [
          { text: '菜单1', click: mockFn2 },
          { text: '菜单2', click: mockFn3 },
        ],
      },
    ]
    const wrapper = mount(Filterbar, { propsData: { filterButtons } })
    const button = wrapper
      .findComponent({ name: 'FilterbarButtons' })
      .findComponent({ name: 'ElButton' })
    const childrenButtons = wrapper
      .findComponent({ name: 'FilterbarButtons' })
      .findAllComponents({ name: 'ElDropdownItem' }).wrappers
    button.trigger('click')
    childrenButtons[0].trigger('click')
    childrenButtons[1].trigger('click')
    expect(mockFn1.call.length).toBe(1)
    expect(mockFn2.call.length).toBe(1)
    expect(mockFn3.call.length).toBe(1)
  })

  it('other render type', () => {
    const filterButtons = [
      () => <div class="function-type">text</div>,
      <div class="jsx-type">text</div>,
    ]
    const wrapper = mount(Filterbar, { propsData: { filterButtons } })
    expect(wrapper.find('div.function-type').exists()).toBe(true)
    expect(wrapper.find('div.jsx-type').exists()).toBe(true)
  })
})

describe('Filter fields', () => {
  it('Field label render', () => {
    const wrapper = mount(Filterbar, {
      propsData: {
        searchButton: false,
        resetButton: false,
        filterFields: [{ type: 'label', label: 'label text' }],
      },
    })
    expect(wrapper.find('div.el-form-item__label').element.innerHTML).toBe(
      'label text'
    )
  })

  it('Fields render', () => {
    const wrapper = mount(Filterbar, {
      propsData: {
        filterFields: [
          { type: 'text', model: 'text' },
          { type: 'number', model: 'number' },
          { type: 'date', model: 'date' },
          { type: 'dateRange', model: 'dateRange' },
          { type: 'timeSelect', model: 'timeSelect' },
          { type: 'timePicker', model: 'timePicker' },
          { type: 'timePickerRange', model: 'timePickerRange' },
          { type: 'dateTime', model: 'dateTime' },
          { type: 'dateTimeRange', model: 'dateTimeRange' },
          { type: 'select', model: 'select' },
          { type: 'cascader', model: 'cascader' },
        ],
      },
    })
    expect(wrapper.findAllComponents({ name: 'FieldText' }).length).toBe(1)
    expect(wrapper.findAllComponents({ name: 'FieldNumber' }).length).toBe(1)
    expect(wrapper.findAllComponents({ name: 'FieldDate' }).length).toBe(1)
    expect(wrapper.findAllComponents({ name: 'FieldDateRange' }).length).toBe(1)
    expect(wrapper.findAllComponents({ name: 'FieldTimeSelect' }).length).toBe(
      1
    )
    expect(wrapper.findAllComponents({ name: 'FieldTimePicker' }).length).toBe(
      1
    )
    expect(
      wrapper.findAllComponents({ name: 'FieldTimePickerRange' }).length
    ).toBe(1)
    expect(wrapper.findAllComponents({ name: 'FieldDateTime' }).length).toBe(1)
    expect(
      wrapper.findAllComponents({ name: 'FieldDateTimeRange' }).length
    ).toBe(1)
    expect(wrapper.findAllComponents({ name: 'FieldSelect' }).length).toBe(1)
    expect(wrapper.findAllComponents({ name: 'FieldCascader' }).length).toBe(1)
  })

  it('Group fields render', () => {
    const wrapper = mount(Filterbar, {
      propsData: {
        filterFields: [
          [
            { type: 'text', model: 'text' },
            { type: 'number', model: 'number' },
          ],
        ],
      },
    })
    expect(wrapper.find('.lv__field-group').exists()).toBe(true)
  })

  it('Filter fields set value', () => {
    const lvStore: any = { filterModel: {} }
    const wrapper = mount(Filterbar, {
      provide: { lvStore },
      propsData: {
        filterFields: [
          { type: 'text', model: 'text' },
          { type: 'number', model: 'number' },
          { type: 'date', model: 'date' },
          { type: 'dateRange', model: 'dateRange' },
          { type: 'timeSelect', model: 'timeSelect' },
          { type: 'timePicker', model: 'timePicker' },
          { type: 'timePickerRange', model: 'timePickerRange' },
          { type: 'dateTime', model: 'dateTime' },
          { type: 'dateTimeRange', model: 'dateTimeRange' },
          { type: 'select', model: 'select' },
          { type: 'cascader', model: 'cascader' },
        ],
      },
    })
    const _findVm = (name: string): any => wrapper.findComponent({ name }).vm
    _findVm('FieldText').value = 'text'
    _findVm('FieldNumber').value = 9527
    _findVm('FieldDate').value = DATE1
    _findVm('FieldDateRange').value = [DATE1, DATE2]
    _findVm('FieldTimeSelect').value = '10:00'
    _findVm('FieldTimePicker').value = DATE1
    _findVm('FieldTimePickerRange').value = [DATE1, DATE2]
    _findVm('FieldDateTime').value = DATE1
    _findVm('FieldDateTimeRange').value = [DATE1, DATE2]
    _findVm('FieldSelect').value = 'option1'
    _findVm('FieldCascader').value = [1, 2, 3, 4]
    expect(lvStore.filterModel).toStrictEqual({
      text: 'text',
      number: 9527,
      date: DATE1,
      dateRange: [DATE1, DATE2],
      timeSelect: '10:00',
      timePicker: DATE1,
      timePickerRange: [DATE1, DATE2],
      dateTime: DATE1,
      dateTimeRange: [DATE1, DATE2],
      select: 'option1',
      cascader: [1, 2, 3, 4],
    })
  })

  it('other render type', () => {
    const filterFields = [
      {
        label: 'label',
        model: 'text',
        render: () => <input class="object-type" />,
      },
      () => <input class="function-type" />,
      <input class="vnode-type" />,
    ]
    const wrapper = mount(Filterbar, { propsData: { filterFields } })
    expect(wrapper.find('.lv__field input.object-type').exists()).toBe(true)
    expect(wrapper.find('.lv__field input.function-type').exists()).toBe(true)
    expect(wrapper.find('.lv__field input.vnode-type').exists()).toBe(true)
  })
})

describe('Filter fields default value', () => {
  const lvStore = {
    filterModel: {
      text: 'text',
      number: 9527,
      date: DATE1,
      dateRange: [DATE1, DATE2],
      timeSelect: '10:00',
      timePicker: DATE1,
      timePickerRange: [DATE1, DATE2],
      dateTime: DATE1,
      dateTimeRange: [DATE1, DATE2],
      select: 'option1',
      cascader: [1, 2, 3, 4],
    },
  }
  const wrapper = mount(Filterbar, {
    provide: { lvStore },
    propsData: {
      filterFields: [
        { type: 'text', model: 'text' },
        { type: 'number', model: 'number' },
        { type: 'date', model: 'date' },
        { type: 'dateRange', model: 'dateRange' },
        { type: 'timeSelect', model: 'timeSelect' },
        { type: 'timePicker', model: 'timePicker' },
        { type: 'timePickerRange', model: 'timePickerRange' },
        { type: 'dateTime', model: 'dateTime' },
        { type: 'dateTimeRange', model: 'dateTimeRange' },
        { type: 'select', model: 'select' },
        { type: 'cascader', model: 'cascader' },
      ],
    },
  })
  const _findVm = (name: string): any => wrapper.findComponent({ name }).vm
  expect(_findVm('FieldText').value).toBe('text')
  expect(_findVm('FieldNumber').value).toBe(9527)
  expect(_findVm('FieldDate').value).toBe(DATE1)
  expect(_findVm('FieldDateRange').value).toEqual([DATE1, DATE2])
  expect(_findVm('FieldTimeSelect').value).toBe('10:00')
  expect(_findVm('FieldTimePicker').value).toBe(DATE1)
  expect(_findVm('FieldTimePickerRange').value).toEqual([DATE1, DATE2])
  expect(_findVm('FieldDateTime').value).toBe(DATE1)
  expect(_findVm('FieldDateTimeRange').value).toEqual([DATE1, DATE2])
  expect(_findVm('FieldSelect').value).toBe('option1')
  expect(_findVm('FieldCascader').value).toEqual([1, 2, 3, 4])
})

describe('Filter fields options resolve', () => {
  const options = [
    { label: 'option1', value: 'option1' },
    { label: 'option2', value: 'option2' },
  ]

  it('array', async () => {
    const wrapper = mount(Filterbar, {
      propsData: {
        filterFields: [
          { type: 'select', model: 'select', options },
          { type: 'cascader', model: 'cascader', options },
        ],
      },
    })
    await wait()
    expect(wrapper.findAllComponents({ name: 'ElOption' }).length).toBe(
      options.length
    )
    expect(
      wrapper.findComponent({ name: 'ElCascader' }).vm.$props.options
    ).toEqual(options)
  })

  it('promise', async () => {
    const wrapper = mount(Filterbar, {
      propsData: {
        filterFields: [
          {
            type: 'select',
            model: 'select',
            options: Promise.resolve(options),
          },
          {
            type: 'cascader',
            model: 'cascader',
            options: Promise.resolve(options),
          },
        ],
      },
    })
    await wait()
    expect(wrapper.findAllComponents({ name: 'ElOption' }).length).toBe(
      options.length
    )
    expect(
      wrapper.findComponent({ name: 'ElCascader' }).vm.$props.options
    ).toEqual(options)
  })

  it('function return array', async () => {
    const wrapper = mount(Filterbar, {
      propsData: {
        filterFields: [
          { type: 'select', model: 'select', options: () => options },
          { type: 'cascader', model: 'cascader', options: () => options },
        ],
      },
    })
    await wait()
    expect(wrapper.findAllComponents({ name: 'ElOption' }).length).toBe(
      options.length
    )
    expect(
      wrapper.findComponent({ name: 'ElCascader' }).vm.$props.options
    ).toEqual(options)
  })

  it('function return promise', async () => {
    const wrapper = mount(Filterbar, {
      propsData: {
        filterFields: [
          {
            type: 'select',
            model: 'select',
            options: () => Promise.resolve(options),
          },
          {
            type: 'cascader',
            model: 'cascader',
            options: () => Promise.resolve(options),
          },
        ],
      },
    })
    await wait()
    expect(wrapper.findAllComponents({ name: 'ElOption' }).length).toBe(
      options.length
    )
    expect(
      wrapper.findComponent({ name: 'ElCascader' }).vm.$props.options
    ).toEqual(options)
  })

  it('invalid options', async () => {
    const wrapper = mount(Filterbar, {
      propsData: {
        filterFields: [{ type: 'select', model: 'select', options: 'options' }],
      },
    })
    await wait()
    expect(wrapper.findAllComponents({ name: 'ElOption' }).length).toBe(0)
  })
})
