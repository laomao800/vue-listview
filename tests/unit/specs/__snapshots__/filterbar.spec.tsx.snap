// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter buttons normal 1`] = `
<form class="el-form el-form--inline lv__filterbar lv__filterbar--fold">
  <!---->
  <div class="lv__filterbar-main">
    <!---->
    <div class="lv__filterbar-inner">
      <div class="lv__filterbar-action lv__filterbar-action--nomore lv__filterbar-action--onleft">
        <div class="el-form-item lv__filterbar-action-submit el-form-item--small" style="transform: translateX(0px);">
          <!---->
          <div class="el-form-item__content"> <button type="button" class="el-button el-button--primary el-button--small" text="搜索">
              <!----><i class="el-icon-search"></i><span>
            搜索
          </span>
            </button> <button type="button" class="el-button el-button--default el-button--small" text="重置">
              <!---->
              <!----><span>
            重置
          </span>
            </button>
            <transition-stub name="el-zoom-in-top">
              <!---->
            </transition-stub>
          </div>
        </div>
        <div class="lv__filterbar-action-ext"> <button type="button" class="el-button lv__filterbar-action-more el-button--primary">
            <!----><i class="el-icon-caret-top"></i>
            <!---->
          </button> </div>
      </div>
      <div class="el-form-item lv__filterbar-buttons el-form-item--small">
        <!---->
        <div class="el-form-item__content"><button type="button" class="el-button el-button--default el-button--small">
            <!---->
            <!----><span>default</span>
          </button><button type="button" class="el-button el-button--text el-button--small">
            <!---->
            <!----><span>text</span>
          </button><button type="button" class="el-button el-button--primary el-button--small">
            <!----><i class="el-icon-edit"></i><span>primary</span>
          </button><button type="button" class="el-button el-button--success el-button--small">
            <!----><i class="el-icon-check"></i><span>success</span>
          </button><button type="button" class="el-button el-button--info el-button--small">
            <!----><i class="el-icon-message"></i><span>info</span>
          </button><button type="button" class="el-button el-button--warning el-button--small">
            <!----><i class="el-icon-star-off"></i><span>warning</span>
          </button><button type="button" class="el-button el-button--danger el-button--small">
            <!----><i class="el-icon-delete"></i><span>danger</span>
          </button>
          <transition-stub name="el-zoom-in-top">
            <!---->
          </transition-stub>
        </div>
      </div>
      <!---->
    </div>
    <!---->
  </div>
  <!---->
</form>
`;
